src/hotspot/share/opto/mulnode.cpp

```diff
@@ -2052,99 +2054,72 @@ const Type* RotateRightNode::Value(PhaseGVN* phase) const {
-    if (check_reverse &&
-        (mask->Opcode() == Op_LShift(bt) ||
-         (bt == T_LONG && mask->Opcode() == Op_ConvI2L))) {
-      // try it the other way around
-      return AndIL_shift_and_mask_is_always_zero(phase, mask, shift, bt, false);
-    }
-    return false;
-  }
-  Node* shift2 = shift->in(2);
-  if (shift2 == nullptr) {
-    return false;
-  }
-  const Type* shift2_t = phase->type(shift2);
-  if (!shift2_t->isa_int() || !shift2_t->is_int()->is_con()) {
-    return false;
+    type = phase->type(expr)->isa_int();
   }
 
-  jint shift_con = shift2_t->is_int()->get_con() & ((shift_bt == T_INT ? BitsPerJavaInteger : BitsPerJavaLong) - 1);
-  if ((((jlong)1) << shift_con) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0) {
-    return true;
+  if (expr->Opcode() == Op_LShift(bt)) {
+    const TypeInt* rhs_t = phase->type(expr->in(2))->isa_int();
+    if (rhs_t == nullptr || !rhs_t->is_con()) {
+      return 0;
+    }
+    long shift = rhs_t->get_con();
+    assert(shift >= 0 && shift < type2aelembytes(bt) * BitsPerByte,
+           "shift amount should have been masked already: %ld", shift);
+    return shift;
   }
 
-  return false;
+  return 0;
 }
 
-// Given an expression (AndX (AddX v1 (LShiftX v2 #N)) #M)
-// determine if the AndX must always produce (AndX v1 #M),
-// because the shift (v2<<N) is bitwise disjoint from the mask #M.
-// The X in AndX will be I or L, depending on bt.
-// Specifically, the following cases fold,
+// Checks whether expr is neutral wrt addition under mask, i.e.
+// an expression of the form (AndX (T+expr) mask) can be simplified to (AndX T mask).
+// The X in AndX must be I or L, depending on bt.
+// Specifically, this holds for the following cases,
 // when the shift value N is large enough to zero out
-// all the set positions of the and-mask M.
-//   (AndI (AddI v1 (LShiftI _ #N)) #M) => (AndI v1 #M)
-//   (AndL (AddI v1 (LShiftL _ #N)) #M) => (AndL v1 #M)
-//   (AndL (AddL v1 (ConvI2L (LShiftI _ #N))) #M) => (AndL v1 #M)
+// all the set positions of the and-mask M:
+//   (AndI (LShiftI _ #N) #M)
+//   (AndL (LShiftL _ #N) #M)
+//   (AndL (ConvI2L (LShiftI _ #N)) #M)
+// including equivalent constant operands:
+//   (AndI (ConI (_ << #N)) #M)
+//   (AndL (ConL (_ << #N)) #M)
 // The M and N values must satisfy ((-1 << N) & M) == 0.
-// Because the optimization might work for a non-constant
-// mask M, and because the AddX operands can come in either
-// order, we check for every operand order.
-Node* MulNode::AndIL_add_shift_and_mask(PhaseGVN* phase, BasicType bt) {
+static bool AndIL_is_zero_element_under_mask(const PhaseGVN* phase, const Node* expr, const Node* mask, BasicType bt) {
+  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
+  if (mask_t == nullptr) {
+    return false;
+  }
+
+  jint zeros = AndIL_min_trailing_zeros(phase, expr, bt);
+  return zeros > 0 && ((((jlong)1) << zeros) > mask_t->hi_as_long() && mask_t->lo_as_long() >= 0);
```

Comment:
FWIW I came up with another way to formulate this. I think it makes even more clear how this is all about "expr is shifted to be completely left of the mask", i.e. we're comparing right-most value bit position to left-most mask bit position, all other bits are irrelevant.

This could also inform how to bias the random generator. Let me know what you think.

```suggestion
  jint expr_trailing_zeros = AndIL_min_trailing_zeros(phase, expr, bt);
  if (expr_trailing_zeros == 0) return false;

  const TypeInteger* mask_t = phase->type(mask)->isa_integer(bt);
  if (mask_t == nullptr || mask_t->lo_as_long() <= 0) return false; // mask = 0 handled in MulNode::Value
  jint mask_bit_width = BitsPerLong - count_leading_zeros(mask_t->hi_as_long());

  return expr_trailing_zeros >= mask_bit_width;
```

Reply:
`mask_t->lo_as_long() == 0` does not imply `mask == 0`, though. Other than that I think it is a great suggestion.

Reply:
You're right. Fixed above. I was trying to get around special casing 0 (count_*_zeros don't like them).